providers/implementations/signature/rsa_sig.c

```diff
@@ -686,6 +686,19 @@ static int rsa_verify_recover(void *vprsactx,
 {
     PROV_RSA_CTX *prsactx = (PROV_RSA_CTX *)vprsactx;
     int ret;
+# ifdef FIPS_MODULE
+    size_t rsabits = RSA_bits(prsactx->rsa);
+
+    if (rsabits < 2048) {
+        if (rsabits != 1024
```

Comment:
I dont think this check is needed (or correct)

Reply:
https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-131Ar2.pdf section 3 table 2 and the text on page 10: "RSA: See FIPS 186-2 and FIPS 186-4, which include modulus lengths of 1024, 1280, 1536 and 1792 bits, may continue to be used for signature verification but not signature generation."

Reply:
I'd read that differently to your lab.

It doesn't say: _which include **only** modulus lengths of_

Reply:
Well... I believe that if the intention was not limit to specific lengths, the text would not say anything about particular modulus lengths.

Reply:
We disagree :grin:

Reply:
My interpretation is the same as Pauli's.. 'Includes' is different from 'must be' one of..

Reply:
FIPS 140-3 IG C.F says

> Signature verification. When performing a FIPS 186-4 signature verification, a module may support all RSA modulus sizes no smaller than 2048, as well as a 1024-bit modulus. When performing a FIPS 186-2 signature verification, the module may support all modulus sizes shown in FIPS 186-2; that is, all numbers in the form of 1024+256*s, where s is a nonnegative integer. The use of RSA modulus sizes listed in this paragraph is approved for signature verification within their respective standard (FIPS 186-4 or FIPS 186-2).

also, for some background:

> Several NIST standards and the CMVP Guidance imply various restrictions on the sizes of the RSA moduli approved for use in the RSA signature generation and the RSA signature verification algorithms. FIPS 186-4approves the use of three modulus sizes: 1024, 2048, and 3072 bits. FIPS 186-2 defers to an ANSI standardX9.31, which permits the use of the RSA moduli sizes that can be represented as 1024 + 256*s bits, where s is a non-negative integer. SP 800-131A Rev2 approves for signature generation all RSA modulus sizes no smaller than 2048.

Does that change your interpretation?